issue 95 - CodeEditor setting and getting first VisibleRow

dkrasner opened the pull request:
## Motivation and Context
This PR allows programmatic setting of the first visible editor row either at instantiation or after, as well as WS messaging to the server to indicate user actions which change the first visible row. 

## Approach
The CodeEditor component uses the Ace-native `scrollToRow` and `goToLine` to set the first line from props (see implementation [here](https://github.com/APrioriInvestments/object_database/blob/daniel-issue-95/object_database/web/content/components/CodeEditor.js#L85)), which is passed to the client at instantiation of the CodeEditor cell. Furthermore the `onScrollTop` Ace editor events triggers a callback which send the first and last visible rows over the web-socket (see [here](https://github.com/APrioriInvestments/object_database/blob/daniel-issue-95/object_database/web/content/components/CodeEditor.js#L233)). 

There is also an option to set the first visible row programmatically via a CodeEditor [method](https://github.com/APrioriInvestments/object_database/blob/daniel-issue-95/object_database/web/cells/cells.py#L2878) which uses the data update message type. 

## How Has This Been Tested?
Playground example and web tests have been added. 

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

dkrasner committed:
adding onScroll callback for the Ace Editor sending over first and last visible rows

dkrasner committed:
adding a scroll to line for the codeeditor and playground example

dkrasner committed:
adjusting codeeditor example

dkrasner committed:
fixing first line in CodeEditori

dkrasner committed:
adding test for CodeEditor setting first visible row

dkrasner committed:
adding updateDate method for CodeEditor for setting first visible row

dkrasner committed:
adding a onChangeFirstRow function to CodeEditor, adding a new example to playground and setting up web to test first row changing

dkrasner committed:
removing web test for first visible row change in CodeEditor as there is no DOM interface way to trigger scroll

dkrasner commented:
@braxtonmckee I decided to add a `onFirstRowChange` function to the CodeEditor.__init__ api since a) it's a useful/natural thing to have b) it would allow for easier testing like we talked about, by Slot'ing something there that another cell responds to

I added `CodeEditorFirstVisibleRowChange` example to the playground and you can see the response to messages going back to the server when you scroll. 

However, I wasn't able to get a proper web selenium test going because Ace uses an internal API to handle scrolling, for which we need access to the editor session that is not available through the py-selenium interface (as far as I can tell). I tried various DOM ways of getting scrolling to happen but no luck. Ace specifically does not want you to do that and it shows. Anyhow I don't want to burn more time on this, unless you think otherwise, since we have a pretty visual example of what is happening client -> server.

One thing to note is that we are using the internal Ace.getFirstVisibleRow api which works fine, but I have been able to get to a place where it's off by a row. For example, combinations of expanding the text/code area and scrolling have made this happen. We could probably build something that is independent of their API but maybe best to stick with it for the moment until it becomes an issue.

dkrasner committed:
setting up WS message tester

dkrasner committed:
working version of wsmessage_tester with code editor example

dkrasner committed:
updating the WSTester example; adding a codeEditor playground message test; wirting docs

dkrasner committed:
adding a nicer display of method and argument ran in the WS tester

dkrasner committed:
adding a web/selenium test for setting the first visible CodeEditor row programmatically from the server

braxtonmckee commented on object_database/web/cells/cells.py:
```diff
@@ -2870,6 +2877,28 @@ def onMessage(self, msgFrame):
+            self.exportData["dataInfo"].append(dataInfo)
+
+        self.wasDataUpdated = True
+        self.markDirty()
```
This is a great example of a place where it would be nice to be able to directly send a message to the component from the python cell, instead of forcing a recalc. Something like `self.messageJsComponent(dict(firstVisibleRow=rowNum))`

braxtonmckee commented on object_database/web/cells/cells.py:
```diff
@@ -3302,3 +3332,40 @@ def __init__(self, content):
+    def onMessage(self, msgFrame):
+        if msgFrame["event"] == "click":
+            # Run the method...
+            self.methodToRun(**self.kwargs)
```
useful to include the result of the call in the data info going back?

braxtonmckee commented on object_database/web/content/components/CodeEditor.js:
```diff
@@ -42,7 +43,7 @@ class CodeEditor extends Component {
             this.editor.setTheme("ace/theme/textmate");
             this.editor.session.setMode("ace/mode/python");
-            this.editor.setAutoScrollEditorIntoView(true);
+            // this.editor.setAutoScrollEditorIntoView(true);
```
does this mean that if I hit 'ctrl-G' it won't scroll that line into view? I don't want to alter the behavior of the component during regular editing - I just want to be able to control the edit state when I switch windows

braxtonmckee committed:
CodeEditor uses sessionState slots to store the first visible row.

This is the same mechanism for storing the current selection state,
and means that code editors that have a valid identity path in the
cells tree above them will remember their UI state when we switch
between them automatically.
